aep/general/0004/aep.md

```diff
@@ -0,0 +1,102 @@
+This means that it is possible (and often desirable) for different APIs to use
+the same type name. For example, a Memcache and Redis API would both want to
+use `Instance` as a type name.
+
+When mapping the relationships between APIs and their resources, however, it
+becomes important to have a single, globally-unique type name. Additionally,
+tools such as Kubernetes or GraphQL interact with APIs from multiple providers.
+
+## Guidance
+
+APIs **must** define a resource type for each resource in the API, according to
+the following pattern: `{API Name}/{Type Name}`. The type name:
+
+- **must** Start with an uppercase letter.
+- **must** Only contain alphanumeric characters.
+- **must** Be of the singular form of the noun.
+- **must** Use PascalCase (UpperCamelCase).
+- For Kubernetes, the type name **must** match the [object][] name.
+- For OpenAPI, the type name **must** match the name of the schema representing
+  the object.
+- For Protobuf, the type name **must** match the name of the protobuf message.
+
+### Examples
+
+Examples of resource types include:
+
+- `pubsub.example.com/Topic`
+- `pubsub.example.com/Subscription`
+- `spanner.example.com/Database`
+- `spanner.example.com/Instance`
+- `networking.istio.io/Instance`
+
+### Annotating resource types
+
+APIs **must** annotate the resource types for each resource in the API using
+the [`google.api.resource`][resource] annotation:
+
+```proto
+// A representation of a Pub/Sub topic.
+message Topic {
+  option (google.api.resource) = {
+    type: "pubsub.googleapis.com/Topic"
+    pattern: "projects/{project}/topics/{topic}"
+    singular: "topic"
+    plural: "topics"
+  };
+
+  // Name and other fields...
+}
+```
+
+- Patterns **must** correspond to the [resource path][resource-paths].
+- Pattern variables (the segments within braces) **must** use `snake_case`, and
+  **must not** use an `_id` suffix.
+- Pattern variables **must** conform to the format `[a-z][_a-z0-9]*[a-z0-9]`.
+- Pattern variables **must** be unique within any given pattern. (e.g.
+  `projects/{abc}/topics/{abc}` is invalid; this is usually a natural corollary
+  of collection identifiers being unique within a pattern.)
+- Singular **must** be the lower camel case of the type.
+  - Pattern variables **must** be the singular form of the resource type e.g. a
+    pattern variable representing a `Topic` resource ID is named `{topic}`.
+- Plural **must** be the lower camel case plural of the singular.
+  - Pattern collection identifier segments **must** match the plural of the
+    resource, except in the case of [nested collections][].
+
+#### Pattern uniqueness
+
+If multiple patterns are defined within a resource, the patterns defined must
+not overlap in the set of resource paths that they can match. In other words, a
+resource path may match at most one of the patterns.
+
+Therefore the following two patterns **must not** be defined within the same
```

Comment:
```suggestion
For example, the following two patterns would not be valid for the same
```

`**must not**` in an example feels a bit off to me, so suggesting slightly different phrasing.